pkg/app/pipedv1/cmd/piped/grpcapi/tool_registry.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2024 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package grpcapi
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"sync"
+	"text/template"
+
+	"golang.org/x/sync/singleflight"
+)
+
+type toolRegistry struct {
+	toolsDir       string
+	tmpDir         string
+	installedTools map[string]struct{}
+	mu             sync.Mutex
+	group          singleflight.Group
+}
+
+type templateValues struct {
+	Name    string
+	Version string
+	OutPath string
+	TmpDir  string
+	Arch    string
+	Os      string
+}
+
+func newToolRegistry(toolsDir, tmpDir string) (*toolRegistry, error) {
+	tmpDir, err := os.MkdirTemp(tmpDir, "tool-registry")
```

Comment:
[Ask] Is it enough just to create the tmp dir for the tool installation in the newToolRegistry.

Reply:
Do you mean to call like `os.MkdirTemp("", "tool-registry")` and do not pass the `tmpDir` as the argument?

Reply:
Yes, I mean fixing as newToolRegistry(toolsDir string) and call os.MkdirTemp inside it as you said.

Reply:
I got your point! That's enough.
I'll change the codes.

Reply:
fixed on this commit
https://github.com/pipe-cd/pipecd/pull/5343/commits/f9f7bf041c6ba8227261cdf65ac4cc7118dbf218